### PR TITLE
Implement lw

### PR DESCRIPTION
**Pull request by dannywillems**
NB: next_instruction_pointer must be updated
Implementation is not correct:
- `lw rt, Offset(rs)`: must load the data at address `rs + offset` into `rt`.

### CODE REVIEW

File: optimism/src/mips/interpreter.rs [hunk 1]
```diff
@@ -243,7 +249,17 @@ pub fn interpret_itype<Env: InterpreterEnv>(env: &mut Env, instr: ITypeInstructi
         ITypeInstruction::LoadUpperImmediate => (),
         ITypeInstruction::Load8 => (),
         ITypeInstruction::Load16 => (),
-        ITypeInstruction::Load32 => (),
+        ITypeInstruction::Load32 => {
+            let rt = env.get_instruction_part(InstructionPart::RT);
+            // Values are the first 16bits
```
Comment (querolita) [inline optimism/src/mips/interpreter.rs:254]: First? Do you mean it is given in little endian? ([see](https://max.cs.kzoo.edu/cs230/Resources/MIPS/MachineXL/InstructionFormats.html), where the immediate value seems to be the "last" 16 bits)

File: optimism/src/mips/interpreter.rs [hunk 2]
```diff
@@ -243,7 +249,17 @@ pub fn interpret_itype<Env: InterpreterEnv>(env: &mut Env, instr: ITypeInstructi
...
+            let immediate_value = (env.get_instruction_part(InstructionPart::RD) << 11)
```
Comment (querolita) [inline optimism/src/mips/interpreter.rs:255]: lgtm

File: optimism/src/mips/interpreter.rs [hunk 3]
```diff
@@ -243,7 +249,17 @@ pub fn interpret_itype<Env: InterpreterEnv>(env: &mut Env, instr: ITypeInstructi
...
+                + (env.get_instruction_part(InstructionPart::Shamt) << 6)
+                + (env.get_instruction_part(InstructionPart::Funct));
+            env.overwrite_register_checked(&rt, &immediate_value);
+            env.set_instruction_pointer(env.get_instruction_pointer() + Env::constant(4u32));
```
Comment (querolita) [inline optimism/src/mips/interpreter.rs:259]: So the reason that you increase by 4 the instruction pointer is that instructions take 32 bits in memory (i.e. 4 bytes) and when you read one instruction you move to the next?

**Review by querolita — APPROVED** [review]

File: optimism/src/mips/interpreter.rs (hunk 3 above)
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs:259]: y

File: optimism/src/mips/interpreter.rs (hunk 1 above)
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs:254]: I change in the next PR.

**Comment by dannywillems** [comment]
Merging addi and lui as there are interesting functions for this PR.

**Review by querolita — APPROVED** [review]

File: optimism/src/mips/interpreter.rs [hunk 4]
```diff
@@ -233,17 +311,60 @@ pub fn interpret_itype<Env: InterpreterEnv>(env: &mut Env, instr: ITypeInstructi
...
+        ITypeInstruction::Load32 => {
+            let dest = env.get_instruction_part(InstructionPart::RT);
+            let addr = env.get_instruction_part(InstructionPart::RS);
+            let offset = env.get_immediate();
+            let addr_with_offset = addr.clone() + offset.clone();
+            debug!("lw {}, {}({})", dest.clone(), offset.clone(), addr.clone());
+            // We load 4 bytes, i.e. one word.
+            debug!("{}, {}({})", dest.clone(), offset.clone(), addr.clone());
+            let v0 = env.fetch_memory(&addr_with_offset);
+            let v1 = env.fetch_memory(&(addr_with_offset.clone() + Env::constant(1)));
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs:353]: Yes. The memory is a vector of bytes.

File: optimism/src/mips/interpreter.rs [hunk 5]
```diff
@@ -233,17 +311,60 @@ pub fn interpret_itype<Env: InterpreterEnv>(env: &mut Env, instr: ITypeInstructi
...
         ITypeInstruction::Load16 => (),
-        ITypeInstruction::Load32 => (),
+        ITypeInstruction::Load32 => {
+            let dest = env.get_instruction_part(InstructionPart::RT);
+            let addr = env.get_instruction_part(InstructionPart::RS);
+            let offset = env.get_immediate();
+            let addr_with_offset = addr.clone() + offset.clone();
+            debug!("lw {}, {}({})", dest.clone(), offset.clone(), addr.clone());
+            // We load 4 bytes, i.e. one word.
+            debug!("{}, {}({})", dest.clone(), offset.clone(), addr.clone());
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs:351]: Oopsie.

Comment (dannywillems) [inline optimism/src/mips/interpreter.rs:351]:
```suggestion
```

**Pull request merged**